src/main/java/uk/gov/ons/ctp/common/rest/RestClient.java

```diff
@@ -174,43 +177,51 @@ public RestTemplate getRestTemplate() {
     HttpEntity<P> httpEntity = createHttpEntity(objToSend, headerParams);
     UriComponents uriComponents = createUriComponents(path, queryParams, pathParams);
     ResponseEntity<T> response;
+    String errorMessage = "request failed for the given path";
     try {
       response = restTemplate.exchange(uriComponents.toUri(), method, httpEntity, clazz);
     } catch (HttpStatusCodeException e) {
       // Failure detected. For 4xx and 5xx status codes
-      String errorMessage =
-          method.name()
-              + " request failed for path: '"
-              + uriComponents
-              + "' Status: "
-              + e.getStatusCode()
-              + " ResponseBody: '"
-              + e.getResponseBodyAsString()
-              + "'";
       if (e.getStatusCode() == HttpStatus.NOT_FOUND) {
-        log.warn(errorMessage);
+        errorMessage = "dealing with NOT_FOUND";
+        logging
+            .with("path", path)
+            .with("methodName", method.name())
+            .with("uriComponents", uriComponents)
+            .with("Status", e.getStatusCode())
+            .with("ResponseBody", e.getResponseBodyAsString())
+            .warn(errorMessage);
       } else if (e.getStatusCode() == HttpStatus.TOO_MANY_REQUESTS) {
         // Caller expected to handle this situation
         log.info("Too many requests response on {} for path: {}", method.name(), path);
       } else {
-        log.error(errorMessage, e);
+        logging
+            .with("path", path)
+            .with("methodName", method.name())
+            .with("uriComponents", uriComponents)
+            .with("statusCode", e.getStatusCode())
+            .with("responseBody", e.getResponseBodyAsString())
+            .error(errorMessage);
+        logging.debug(errorMessage, e);
       }
       throw new ResponseStatusException(
           mapToExternalStatus(e.getStatusCode()), e.getResponseBodyAsString(), e);
     } catch (RestClientException e) {
-      log.error(method.name() + " failed for path: '" + uriComponents + "'", e);
+      logging.with("path", path).with("methodName", method.name()).error(errorMessage, e);
       throw new ResponseStatusException(
           HttpStatus.INTERNAL_SERVER_ERROR, "Internal processing error");
     }
 
     T responseObject = response.getBody();
     if (responseObject == null) {
-      String errorMessage =
-          "Empty body returned for path '"
-              + uriComponents.toUriString()
-              + "'. Status code: "
-              + response.getStatusCode();
-      log.error(errorMessage);
+      errorMessage = "Empty body returned for given path";
+      logging
+          .with("path", path)
+          .with("methodName", method.name())
+          .with("uriComponents", uriComponents)
+          .with("statusCode", response.getStatusCode())
+          .with("responseBody", response.getBody())
+          .error(errorMessage);
```

Comment:
Hold on a minute , I thought we all agreed that when we logged errors, that we would include the stack trace for diagnostics.

Reply:
yes @rpcatling we did but please see Phil's comments about not populating prod with stacktraces..which is a little worrying because the logging in rhsvc and ccsvc, as we agreed have been updated already to include stacktrace for errors and doesn't match to what Phil says